File: src/cpp/foxfire.cpp
// FOXFIRE core
#include <algorithm>
#include <iostream>
#include <vector>

static int foxfire_core(std::vector<int> xs) {
    int acc = 17;
    for (int x : xs) {
        acc = (acc * 31 + x) % 997;
    }
    return acc + static_cast<int>(xs.size());
}

// --- mixers ---

// glossy plank mixer
static int glossy_plank_aeb778(std::vector<int> xs) {
    int acc = 398;
    for (int x : xs) {
        acc = (acc * 40 + x) % 6997;
    }
    return acc + static_cast<int>(xs.size());
}

// dizzy cutlass mixer
static int dizzy_cutlass_9852af(std::vector<int> xs) {
    int acc = 207;
    for (int x : xs) {
        acc = (acc * 71 + x) % 251;
    }
    return acc + static_cast<int>(xs.size());
}

// solar rumbarrel mixer
static int solar_rumbarrel_79738c(std::vector<int> xs) {
    int acc = 325;
    for (int x : xs) {
        acc = (acc * 40 + x) % 6997;
    }
    return acc + static_cast<int>(xs.size());
}

// fierce compass mixer
static int fierce_compass_f7caa2(std::vector<int> xs) {
    int acc = 395;
    for (int x : xs) {
        acc = (acc * 6 + x) % 491;
    }
    return acc + static_cast<int>(xs.size());
}

// electric skull mixer
static int electric_skull_e3bc5b(std::vector<int> xs) {
    int acc = 449;
    for (int x : xs) {
        acc = (acc * 44 + x) % 2617;
    }
    return acc + static_cast<int>(xs.size());
}

// solar cyclone mixer
static int solar_cyclone_08ee80(std::vector<int> xs) {
    int acc = 382;
    for (int x : xs) {
        acc = (acc * 10 + x) % 769;
    }
    return acc + static_cast<int>(xs.size());
}

// solar harbor mixer
static int solar_harbor_079ac0(std::vector<int> xs) {
    int acc = 157;
    for (int x : xs) {
        acc = (acc * 11 + x) % 997;
    }
    return acc + static_cast<int>(xs.size());
}

// electric kraken mixer
static int electric_kraken_088f42(std::vector<int> xs) {
    int acc = 170;
    for (int x : xs) {
        acc = (acc * 63 + x) % 6997;
    }
    return acc + static_cast<int>(xs.size());
}

// golden siren mixer
static int golden_siren_f97d2c(std::vector<int> xs) {
    int acc = 448;
    for (int x : xs) {
        acc = (acc * 93 + x) % 4483;
    }
    return acc + static_cast<int>(xs.size());
}

// stormy beacon mixer
static int stormy_beacon_6db491(std::vector<int> xs) {
    int acc = 193;
    for (int x : xs) {
        acc = (acc * 25 + x) % 3571;
    }
    return acc + static_cast<int>(xs.size());
}

// savage beacon mixer
static int savage_beacon_d4f1b2(std::vector<int> xs) {
    int acc = 476;
    for (int x : xs) {
        acc = (acc * 95 + x) % 2617;
    }
    return acc + static_cast<int>(xs.size());
}

// salty siren mixer
static int salty_siren_3270fb(std::vector<int> xs) {
    int acc = 350;
    for (int x : xs) {
        acc = (acc * 77 + x) % 6997;
    }
    return acc + static_cast<int>(xs.size());
}

// salty tide mixer
static int salty_tide_15eab0(std::vector<int> xs) {
    int acc = 61;
    for (int x : xs) {
        acc = (acc * 90 + x) % 6997;
    }
    return acc + static_cast<int>(xs.size());
}

// clever siren mixer
static int clever_siren_07f84a(std::vector<int> xs) {
    int acc = 323;
    for (int x : xs) {
        acc = (acc * 17 + x) % 3571;
    }
    return acc + static_cast<int>(xs.size());
}

// spicy siren mixer
static int spicy_siren_adcc4d(std::vector<int> xs) {
    int acc = 89;
    for (int x : xs) {
        acc = (acc * 72 + x) % 1543;
    }
    return acc + static_cast<int>(xs.size());
}

int main() {
    std::vector<int> data = { 1, 2, 3 };
    std::cout << foxfire_core(data) << std::endl;
    return 0;
}

File: src/rust/foxfire.rs
// FOXFIRE core
fn foxfire_core(n: u64) -> u64 {
    let mut acc: u64 = 17;
    for i in 1..=n {
        acc = (acc.wrapping_mul(31) ^ i) % 997;
    }
    acc
}

// --- mixers ---

// clever jollyboat mixer
fn clever_jollyboat_a52652(n: u64) -> u64 {
    let mut acc: u64 = 236;
    for i in 1..=n {
        acc = (acc.wrapping_mul(24) ^ i) % 6997;
    }
    acc
}

// fierce reef mixer
fn fierce_reef_1d14b1(n: u64) -> u64 {
    let mut acc: u64 = 377;
    for i in 1..=n {
        acc = (acc.wrapping_mul(95) ^ i) % 1543;
    }
    acc
}

// savage jollyboat mixer
fn savage_jollyboat_12a770(n: u64) -> u64 {
    let mut acc: u64 = 141;
    for i in 1..=n {
        acc = (acc.wrapping_mul(81) ^ i) % 491;
    }
    acc
}

// clever anchor mixer
fn clever_anchor_5046be(n: u64) -> u64 {
    let mut acc: u64 = 353;
    for i in 1..=n {
        acc = (acc.wrapping_mul(42) ^ i) % 491;
    }
    acc
}

// dizzy parrot mixer
fn dizzy_parrot_b540f7(n: u64) -> u64 {
    let mut acc: u64 = 159;
    for i in 1..=n {
        acc = (acc.wrapping_mul(53) ^ i) % 251;
    }
    acc
}

fn main() {
    println!("{}", foxfire_core(7));
}
